### PR TITLE
doc: ARCHITECTURE.md a rough map for developers

### PR DESCRIPTION
**Pull request by coryan**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5834)

### CODE REVIEW

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+- Bigtable has a [bigtable::Table](/google/cloud/bigtable/table.h) client, as
+  well as [TableAdmin](/google/cloud/bigtable/table_admin.h) and
+  [InstanceAdmin](/google/cloud/bigtable/instance_admin.h) clients.
+- Storage has a single [storage::Client](/google/cloud/storage/client.h) client
+  class.
+- Pub/Sub has two non-admin clients, the
+  [pubsub::Publisher](/google/cloud/pubsub/publisher.h)
+  and the [pubsub::Subscriber](/google/cloud/pubsub/subscriber.h).
+- Spanner has one non-admin client, the
```
Comment (devjgm) [inline ARCHITECTURE.md:106]: Spanner also has the `DatabaseAdminClient` and `InstanceAdminClient`. Did you want to mention those here too?

**Review by devjgm — APPROVED** [review]

File: ARCHITECTURE.md (hunk 1 above)
Comment (coryan) [reply ARCHITECTURE.md:109]: Done.

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  latency, throughput, and CPU overhead (as in "low CPU overhead") requirements.
+- `ci/` may contain helper scripts used in the CI builds, for example, to
+  startup and shutdown any emulators used by the library.
+- `doc/` contains the landing page for any Doxygen documentation, as well as
+  additional Doxygen docs that are not extracted from the code comments.
+- `integration_tests/` (or simply `tests/` in some libraries) contain any
+  integration tests where we test the library using emulators and/or the
+  production environment.
+- `testing/` contains testing utilities shared by the unit tests and the
+  integration tests of the library. They are not intended to be part of the
```
Comment (devbww) [inline ARCHITECTURE.md:84]: "part of the \<something>"?

File: ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+## The `*Client` classes
+
+The main interface for each library is the `${library}::*Client` class. Roughly
+speaking each "client" corresponds to a `service` in the `.proto` definitions
+for the service. It is common for services to have separate `service`
+definitions for "admin" operations, such as creating a new instance of the
+service, or changing permissions, as opposed to "normal" operations, such as
+inserting a new row, or publishing a new message. When "admin" services are
+defined, there are separate `${library}::Client` objects.
```
Comment (devbww) [inline ARCHITECTURE.md:94]: Did you mean the previous name to be `${library}::Client` and this one the `${library}::*Client` glob?

File: ARCHITECTURE.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+## What these libraries do
+
+The goal of the libraries is to provide idiomatic C++ libraries to access
+services in [Google Cloud Platform](https://cloud.google.com). All services are
+in scope, though only a handful of services have been implemented to this point.
+
+What do we mean by idiomatic? We mean that C++ developers will find the
+APIs familiar, or "natural", that these APIs will fit well with the rest of the
+C++ ecosystem, that very few new "concepts" are needed to understand how to use
```
Comment (devbww) [inline ARCHITECTURE.md:17]: s/that/and that/

File: ARCHITECTURE.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+these libraries.
+
+More specifically, some of the functionality offered by these libraries include:
+
+- The libraries automatically *retry* RPCs that fail with a retriable error,
+  such as network failures.
+- The libraries only retry operations when it is safe to do so, "delete the
```
Comment (devbww) [inline ARCHITECTURE.md:24]: s/so, "delete/so. "Delete/

File: ARCHITECTURE.md [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+  first row" is not a safe operation to retry when the previous result is
+  unknown. In the documentation we refer to operations that are safe to retry
+  as [idempotent](https://en.wikipedia.org/wiki/Idempotence).
+- The libraries also *resume* large downloads or database reads from the last
+  fully received element, minimizing the amount of duplicate data received by
+  your application.
+- The application can always change the retry and resume policies, the libraries
+  have reasonable defaults, and implement best practices for backing off before
```
Comment (devbww) [inline ARCHITECTURE.md:32]: Perhaps "The libraries have reasonable defaults for retry and resume policies, and implement ...".

File: ARCHITECTURE.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+  [pubsub::TopicAdminClient](/google/cloud/pubsub/topic_admin_client.h),
+  and [pubsub::SubscriptionAdminClient](/google/cloud/pubsub/subscription_admin_client.h).
+- Spanner has one non-admin client, the
+  [spanner::Client](/google/cloud/spanner/client.h). It also has two admin
+  clients:
+  [pubsub::InstanceAdminClient](/google/cloud/spanner/instance_admin_client.h),
+  and
+  [pubsub::DatabaseAdminClient](/google/cloud/spanner/database_admin_client.h).
+
+Generally these classes are very "thin", they take function arguments from the
```
Comment (devbww) [inline ARCHITECTURE.md:116]: s/,/;/

File: ARCHITECTURE.md [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+[Decorators](https://en.wikipedia.org/wiki/Decorator_pattern), which simplifies
+their testing.
+
+| Layer | Description |
+| ----- | ----------- |
+| `*Logging` | Optional `*Stub` decorator that logs each request and response |
+| `*Metadata` | Injects resource metadata headers for routing |
+| `*RoundRobin` | Round-robins over several `*Stub`s, not all libraries have them |
+
+## Exceptions
```
Comment (devbww) [inline ARCHITECTURE.md:164]: It might be better to use a synonym for "exceptions" when talking about a C++ library.

**Review by devbww — COMMENTED** [review]
Late, but hopefully helpful.

**Comment by coryan** [comment]
These are helpful, I have sent a new PR (#5837), but will respond here as a checklist for myself.

File: ARCHITECTURE.md (hunk 8 above)
Comment (coryan) [reply ARCHITECTURE.md:164]: Done.

File: ARCHITECTURE.md (hunk 7 above)
Comment (coryan) [reply ARCHITECTURE.md:116]: Done.

File: ARCHITECTURE.md (hunk 6 above)
Comment (coryan) [reply ARCHITECTURE.md:32]: Took a stab at it.

File: ARCHITECTURE.md (hunk 5 above)
Comment (coryan) [reply ARCHITECTURE.md:24]: Changed.

File: ARCHITECTURE.md (hunk 4 above)
Comment (coryan) [reply ARCHITECTURE.md:17]: Done.

File: ARCHITECTURE.md (hunk 3 above)
Comment (coryan) [reply ARCHITECTURE.md:94]: Fixed, thanks.

File: ARCHITECTURE.md (hunk 2 above)
Comment (coryan) [reply ARCHITECTURE.md:84]: "public API", but this is implied in the heading, so I am removing it.

**Pull request merged**